File: implement-strStr-1.java
class Solution {
    public int strStr(String haystack, String needle) {
        int size = needle.length();
        
        for (int i = 0; i <= haystack.length()-size; i++) {
            if (equals(haystack, i, i+size, needle))
                return i;
        }
        
        return -1;
    }
    
    private boolean equals(String haystack, int start, int end, String needle) {
        int index = 0;
        
        for (int i = start; i < end; i++) {
            if (haystack.charAt(i) != needle.charAt(index++))
                return false;
        }
        
        return true;
    }
}

File: implement-strStr-1.scala
object Solution {
    def strStr(haystack: String, needle: String): Int = {
        strStr(haystack, 0, needle)
    }
    
    def strStr(haystack: String, start: Int, needle: String): Int = {
        val size = needle.length()
        if (haystack.length() < start + size)
            return -1
        
        if (haystack.substring(start, start + size).equals(needle))
            return start
        
        strStr(haystack, start + 1, needle);
    }
}
